File: Bwl.Imaging/Matrix/IntPtr/AutoPinner.vb
Imports System.Runtime.InteropServices

'Using ap As New AutoPinner(MyManagedObject)
'   UnmanagedIntPtr = ap ' Use the operator to retrieve the IntPtr
'   'do your stuff
'End Using

Friend Class AutoPinner
    Implements IDisposable

    Private _pinnedArray As GCHandle

    Public Sub New(ByVal obj As Object)
        _pinnedArray = GCHandle.Alloc(obj, GCHandleType.Pinned)
    End Sub

    Public Function GetIntPtr() As IntPtr
        Return _pinnedArray.AddrOfPinnedObject()
    End Function

    Public Sub Dispose() Implements IDisposable.Dispose
        _pinnedArray.Free()
    End Sub
End Class

File: Bwl.Imaging/Matrix/IntPtr/AutoPinner.cs
using System;
using System.Runtime.InteropServices;

namespace Bwl.Imaging
{

    // Using ap As New AutoPinner(MyManagedObject)
    // UnmanagedIntPtr = ap ' Use the operator to retrieve the IntPtr
    // 'do your stuff
    // End Using

    internal class AutoPinner : IDisposable
    {

        private GCHandle _pinnedArray;

        public AutoPinner(object obj)
        {
            _pinnedArray = GCHandle.Alloc(obj, GCHandleType.Pinned);
        }

        public IntPtr GetIntPtr()
        {
            return _pinnedArray.AddrOfPinnedObject();
        }

        public void Dispose()
        {
            _pinnedArray.Free();
        }
    }
}